File: cpptet_v0.5/CTetris.cpp
#include "CTetris.h"

extern Matrix **(*setOfCBlockObjects);
extern Matrix **(*setOfBlockObjects);

extern int nBlockDegrees;
extern int nBlockTypes;
extern int iScreenDw;

void init(int **setOfCBlockArrays, int MAX_BLK_TYPES, int MAX_BLK_DEGREES){
    int i = 0; int j = 0;
    
    tetinit(setOfCBlockArrays, MAX_BLK_TYPES, MAX_BLK_DEGREES);

    setOfCBlockObjects = new Matrix**[nBlockTypes];
    for (int y = 0; y < nBlockTypes; y++)
        setOfCBlockObjects[y] = new Matrix*[nBlockDegrees];
    for (int y = 0; y < nBlockTypes; y++)
        for (int x = 0; x < nBlockDegrees; x++)
            setOfCBlockObjects[y][x] = nullptr;

    for (i=0; i < nBlockTypes; i++)
    {
        for (j=0; j < nBlockDegrees; j++){
            int *arrptr = setOfCBlockArrays[i*4 + j];
            int arrsize = 0;
            while(1){
                if (arrptr[arrsize] == -1){
                    break;
                }
                else{
                    arrsize++;
                }
            }
            Matrix *obj;
            obj = new Matrix(arrptr, sqrt(arrsize), sqrt(arrsize));
            obj->mulc(i+1);
            cout << *obj <<endl;
            setOfCBlockObjects[i][j] = obj;
        }
    }

    return;
}

TetrisState CTetris::accept(char key){
    TetrisState state;

    if ((key >= '0') && (key <= '6')){
        if (justStarted == false){
            this->deleteFullLines();
        }
        iCScreen = new Matrix(oCScreen);
    }
    
    state = tetaccept(key);

    currCBlk = setOfCBlockObjects[this->idxBlockType][this->idxBlockDegree];
    tempBlk = iCScreen->clip(this->top, 
                             this->left, 
                             this->top + currCBlk->get_dy(), 
                             this->left + currCBlk->get_dx());

    tempBlk = tempBlk->add(currCBlk);

    oCScreen = new Matrix(iCScreen);
    oCScreen->paste(tempBlk, this->top, this->left);

    return state;
}

void CTetris::deleteFullLines(){
    system("clear");
    std::cout << "call CTetris::deleteFullLines()" << std::endl;
    bool is_zero = false;
    int ** array = oScreen->get_array();

    int dx = oScreen->get_dx() - iScreenDw*2;

    for (int y=0; y < oScreen->get_dy() - iScreenDw; y++){
        for (int x=0; x < dx; x++){
            if (array[y][x] == 0){
                is_zero = true;
            }
        }

        if (!is_zero){ // if 0 not in array[y]
            for (int x=0; x < dx; x++){
                array[y][x] = 0;
            }
            top = 1;
            left = 0;
            currCBlk = oScreen->clip(0, 0, y, oScreen->get_dx());
            tempBlk = iScreen->clip(0, 0, y, oScreen->get_dx());
            tempBlk->paste(currCBlk, 0, 0);
            oScreen = new Matrix(iScreen);
            oScreen->paste(tempBlk, top, left);
        }
    }
}

CTetris::CTetris(int cy, int cx) : Tetris(cy, cx){
    int **arrayScreen2D;
    arrayScreen2D = createArrayScreen();

    int k = 0;
    for (int y=0; y < arrayScreenDy; y++){   
        for (int x=0; x < arrayScreenDx; x++){
            arrayScreen[k] = arrayScreen2D[y][x];
            k++;
        }
    }

    this->iCScreen = new Matrix(arrayScreen, arrayScreenDy, arrayScreenDx);
    this->oCScreen = new Matrix(iCScreen);
}

File: pytet_v0.4/ctetris.py
from matrix import *
from tetris import *
from enum import Enum

class TetrisState(Enum):
    Running = 0
    NewBlock = 1
    Finished = 2
### end of class TetrisState():

class CTetris(Tetris):
    nBlockTypes = 0
    nBlockDegrees = 0
    setOfBlockObjects = 0
    iScreenDw = 0   # larget enough to cover the largest block

    def __init__(self, iScreenDy, iScreenDx):
        Tetris.__init__(self, iScreenDy, iScreenDx)
        return

    def accept(self, key):
        self.state = TetrisState.Running

        if key >= '0' and key <= '6':
            if self.justStarted == False:
                self.deleteFullLines()
            self.iScreen = Matrix(self.oScreen)
            self.idxBlockType = int(key)
            self.idxBlockDegree = 0
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            self.top = 0
            self.left = Tetris.iScreenDw + self.iScreenDx//2 - self.currBlk.get_dx()//2
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            binaryBlk = self.tempBlk.binary() + self.currBlk.binary()
            self.tempBlk = self.tempBlk + self.currBlk
            self.justStarted = False
            print()

            if binaryBlk.anyGreaterThan(1):
            #if self.tempBlk.anyGreaterThan(1):
                self.state = TetrisState.Finished
            self.oScreen = Matrix(self.iScreen)
            self.oScreen.paste(self.tempBlk, self.top, self.left)
            return self.state
        elif key == 'q':
            pass
        elif key == 'a': # move left
            self.left -= 1
        elif key == 'd': # move right
            self.left += 1
        elif key == 's': # move down
            self.top += 1
        elif key == 'w': # rotate the block clockwise
            self.idxBlockDegree = (self.idxBlockDegree + 1) % Tetris.nBlockDegrees
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
        elif key == ' ': # drop the block
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            binaryBlk = self.tempBlk.binary() + self.currBlk.binary()
            while not binaryBlk.anyGreaterThan(1):
                self.top += 1
                self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
                binaryBlk = self.tempBlk.binary() + self.currBlk.binary()
                self.tempBlk = self.tempBlk + self.currBlk
        else:
            print('Wrong key!!!')
            
        self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
        binaryBlk = self.tempBlk.binary() + self.currBlk.binary()
        self.tempBlk = self.tempBlk + self.currBlk

        #if self.tempBlk.anyGreaterThan(1):   ## 벽 충돌시 undo 수행
        if binaryBlk.anyGreaterThan(1): 
            if key == 'a': # undo: move right
                self.left += 1
            elif key == 'd': # undo: move left
                self.left -= 1
            elif key == 's': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            elif key == 'w': # undo: rotate the block counter-clockwise
                self.idxBlockDegree = (self.idxBlockDegree - 1) % Tetris.nBlockDegrees
                self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            elif key == ' ': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            binaryBlk = self.tempBlk.binary() + self.currBlk.binary()
            self.tempBlk = self.tempBlk + self.currBlk

        self.oScreen = Matrix(self.iScreen)
        self.oScreen.paste(self.tempBlk, self.top, self.left)

        return self.state

    def deleteFullLines(self):
        array = self.oScreen.get_array()
        
        for y in range(self.oScreen.get_dy()-Tetris.iScreenDw):
            if 0 not in array[y]:
                # delete line
                del_line = array[y]
                array[y] = [0]*len(array[y])
                del del_line

                # move
                self.top = 1
                self.left = 0

                self.currBlk = self.oScreen.clip(0, 0, y, self.oScreen.get_dx())
                self.tempBlk = self.iScreen.clip(0, 0, y, self.oScreen.get_dx())
                self.tempBlk.paste(self.currBlk, 0, 0)
                self.oScreen = Matrix(self.iScreen)
                self.oScreen.paste(self.tempBlk, self.top, self.left)
                
### end of class Tetris():
